### PR TITLE
Flatten configuration and match Dropwizard names

### PR DESCRIPTION
**Pull request by nmiyake**

### CODE REVIEW

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java [hunk 1]
```diff
@@ -28,18 +30,44 @@
...
+
+    @SuppressWarnings("checkstyle:designforextension")
+    @Value.Default
+    public String keyStoreType() {
+        return "JKS";
+    }
+
+    // alias of the certificate to use in the key store.
+    // If absent, first entry returned by key store is used.
+    public abstract Optional<String> certAlias();
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: Even though this is what Dropwizard uses I think it's ambguous here and would prefer something appropriately specific, like keyStoreKeyAlias

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java [hunk 2]
```diff
@@ -28,18 +30,44 @@
...
+
+    @Value.Check
+    protected final void check() {
+        Preconditions.checkArgument(!keyStorePath().isPresent() || keyStorePassword().isPresent(),
+                "keyStorePassword cannot be absent if keyStorePath is present");
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: You need to also reject if keyStorePassword is present but keyStorePath is not

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java (hunk 1 above)
Comment (nmiyake) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: Yeah in this case clarity probably should win out -- will update.

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java [hunk 3]
```diff
@@ -28,18 +30,44 @@
 @Value.Style(visibility = ImplementationVisibility.PACKAGE)
 public abstract class SslConfiguration {
 
-    @Value.Parameter
-    public abstract TrustStoreConfiguration trust();
+    public abstract URI trustStorePath();
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: These are different from the Dropwizard spec in that they take URIs and not paths. Is this still actually a reasonable name?

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java (hunk 2 above)
Comment (nmiyake) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: Noticed right after I pushed this and implemented locally -- pushing up now :)

```
    @Value.Check
    protected final void check() {
        Preconditions.checkArgument(keyStorePath().isPresent() == keyStorePassword().isPresent(),
                "keyStorePath and keyStorePassword must both be present or both be absent");
    }
```

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java (hunk 3 above)
Comment (nmiyake) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: Fair point. I'm okay with `trustStoreUri` and `keyStoreUri` to keep same general format but to make the type more clear from the param name.

Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: I guess it's up to what others want. I don't have strong preferences, but I think this does remove clarity.

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java [hunk 4]
```diff
@@ -28,18 +30,51 @@
...
+
+    @Value.Check
+    protected final void check() {
+        Preconditions.checkArgument(
+                keyStorePath().isPresent() == keyStorePassword().isPresent(),
+                "keyStorePath and keyStorePassword must both be present or both be absent");
+
+        Preconditions.checkArgument(
+                !keyStoreKeyAlias().isPresent() || keyStorePath().isPresent(),
+                "keyStoreKeyAlias must be present if keyStorePath is present");
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: hm, I think `keyStoreKeyAlias` may only be present when `keyStorePath` is present, but is not required.

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslSocketFactories.java [hunk 5]
```diff
@@ -75,23 +76,24 @@ private static TrustManagerFactory createTrustManagerFactory(TrustStoreConfigura
         }
     }
 
-    private static KeyManagerFactory createKeyManagerFactory(KeyStoreConfiguration keyStoreConfig) {
+    private static KeyManagerFactory createKeyManagerFactory(SslConfiguration sslConfig) {
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslSocketFactories.java]: the args to this should probably explicitly be `Path keyStorePath`, `String keyStorePassword`, `String keyStoreType`, `Optional<String> keyStoreKeyAlias`

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslSocketFactories.java [hunk 6]
```diff
@@ -59,12 +60,12 @@ public static SSLContext createSslContext(SslConfiguration config) {
         }
     }
 
-    private static TrustManagerFactory createTrustManagerFactory(TrustStoreConfiguration trustStoreConfig) {
+    private static TrustManagerFactory createTrustManagerFactory(SslConfiguration sslConfig) {
```
Comment (markelliot) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslSocketFactories.java]: the args to this should probably explicitly be `Path trustStorePath`, `Optional<String> trustStorePassword`, `String trustStoreType`

File: ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java (hunk 4 above)
Comment (nmiyake) [inline ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java]: Oops, sorry, message is incorrect here. What you're saying is correct -- this is meant to throw an error if `keyStoreKeyAlias` is specified when `keyStorePath` is not present.

**Comment by markelliot** [comment]
this lgtm

**Comment by nmiyake** [comment]
Great -- will rebase, squash, and push back up

**Pull request merged**